File: code/csharp/CdkWorkshop/Constructs/HitCounter.cs
using System.Collections.Generic;
using Amazon.CDK;
using Amazon.CDK.AWS.DynamoDB;
using Amazon.CDK.AWS.Lambda;

namespace CdkWorkshop.Constructs
{
    public class HitCounterProps
    {
        public IIFunction Downstream { get; set; }
    }

    public class HitCounter : Construct
    {
        public IIFunction Handler { get; }

        public HitCounter(Construct scope, string id, HitCounterProps props) : base(scope, id)
        {
            var table = new Table(this, "Hits", new TableProps {
                PartitionKey = new Attribute {
                    Name = "path",
                    Type = AttributeType.STRING,
                }
            });

            Handler = new Function(this, "HitConterHandler", new FunctionProps
            {
                Runtime = Runtime.DOTNET_CORE_2_1,
                Timeout = Duration.Seconds(30),
                Code = Code.Asset("./HitCounterFunction/src/HitCounterFunction/bin/Debug/netcoreapp2.1"),
                Handler = "HitCounterFunction::HitCounterFunction.Function::FunctionHandler",
                Environment = new Dictionary<string, object>
                {
                    {"DOWNSTREAM_FUNCTION_NAME", props.Downstream.FunctionName},
                    {"HITS_TABLE_NAME", table.TableName}
                }
            });

            table.GrantReadWriteData(Handler.Role);
            props.Downstream.GrantInvoke(Handler.Role);
        }
    }
}

File: code/typescript/main-workshop/lib/hitcounter.ts
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as dynamodb from 'aws-cdk-lib/aws-dynamodb';
import { Construct } from 'constructs';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import * as path from 'path';

export interface HitCounterProps {
  /** the function for which we want to count url hits **/
  downstream: lambda.IFunction;
}

export class HitCounter extends Construct {
  /** allows accessing the counter function */
  public readonly handler: lambda.Function;

  /** the hit counter table */
  public readonly table: dynamodb.Table;

  constructor(scope: Construct, id: string, props: HitCounterProps) {
    super(scope, id);

    const table = new dynamodb.Table(this, 'Hits2', {
      partitionKey: { name: 'path', type: dynamodb.AttributeType.STRING },
      billingMode: dynamodb.BillingMode.PAY_PER_REQUEST
    });
    this.table = table;

    this.handler = new NodejsFunction(this, 'HitCounterHandler', {
      runtime: lambda.Runtime.NODEJS_20_X,
      handler: 'handler',
      entry: path.join(__dirname, '../lambda/hitcounter.ts'),
      environment: {
        DOWNSTREAM_FUNCTION_NAME: props.downstream.functionName,
        HITS_TABLE_NAME: table.tableName
      }
    });

    // grant the lambda role read/write permissions to our table
    table.grantReadWriteData(this.handler);

    // grant the lambda role invoke permissions to the downstream function
    props.downstream.grantInvoke(this.handler);
  }
}
